File: analyze_protein_trajectory/HP36/traj_A/average_ene_vs_residue/average_ene_vs_residue.f
c
c     calculate the average of ene_vs_residue
c
c       input  : list_files_ene_res
c       output : average_ene_vs_residue.dat
c

      program statistics_ene_vs_residue

      implicit none

      integer maxres
      parameter (maxres=1000)

      integer nres, ndata
      real*8  ave_ene_tot(maxres)
      real*8  ave_ene_LJ(maxres)
      real*8  ave_ene_cou(maxres)

      real*8  ave2_ene_tot(maxres), sig_ene_tot

      real*8  ene_res, ene_res_LJ, ene_res_cou
      integer ires, ires_dum

      real*8  ave_ene_all

      character*200 filename

c
c     initialization
c

      ndata = 0

      do ires=1,maxres

        ave_ene_tot(ires) = 0.d0
        ave_ene_LJ(ires)  = 0.d0
        ave_ene_cou(ires) = 0.d0

        ave2_ene_tot(ires) = 0.d0

      enddo

      ave_ene_all = 0.d0

c
c     read data and caluclate the statistics
c

      open(10,file='list_files_ene_res',status='old')

  100 read(10,'(A)',end=190,err=90) filename
      open(20,file=filename,status='old',err=90)

        ndata = ndata + 1

        read(20,*,err=90) nres

        if (nres.gt.maxres)  then
          print *,'error: number of residues exceeds maxres'
          goto 99
        endif

        do ires=1,nres

          read(20,*,err=90) ires_dum,ene_res,ene_res_LJ,ene_res_cou

          if (ires_dum.ne.ires)  then
            print *,'error:  inconsistency in the residue number' 
            goto 99
          endif

          ave_ene_tot(ires) = ave_ene_tot(ires) + ene_res
          ave_ene_LJ(ires)  = ave_ene_LJ(ires)  + ene_res_LJ
          ave_ene_cou(ires) = ave_ene_cou(ires) + ene_res_cou

          ave2_ene_tot(ires) = ave2_ene_tot(ires) + ene_res*ene_res

          ave_ene_all = ave_ene_all + ene_res

        enddo

        close(20,err=90)

      goto 100

  190 continue

      close(10,err=90)

c     normalization

      do ires=1,nres

        ave_ene_tot(ires) = ave_ene_tot(ires) / ndata
        ave_ene_LJ(ires)  = ave_ene_LJ(ires)  / ndata
        ave_ene_cou(ires) = ave_ene_cou(ires) / ndata

        ave2_ene_tot(ires) = ave2_ene_tot(ires) / ndata

      enddo

      ave_ene_all = ave_ene_all / ndata

c
c     save average_ene_vs_residue.dat
c

      open(30,file='average_ene_vs_residue.dat',err=90)

      do ires=1,nres

        sig_ene_tot = dsqrt(ave2_ene_tot(ires)-ave_ene_tot(ires)**2)

        write(30,990) ires, ave_ene_tot(ires),
     .                      ave_ene_LJ(ires),
     .                      ave_ene_cou(ires),
     .                      sig_ene_tot

      enddo

      close(30,err=90)

c
c     print out summary
c

c     on screen

      write(6,*)
      write(6,*) 'total number of data used = ', ndata
      write(6,*)
      write(6,*) 'ave_ene_all         : ', ave_ene_all
      write(6,*)

c     on file

      open(30,file='summary_average_ene_vs_residue.dat',err=90)

      write(30,*)
      write(30,*) 'total number of data used = ', ndata
      write(30,*)
      write(30,*) 'ave_ene_all         : ', ave_ene_all
      write(30,*)

      close(30,err=90)

c
c     format
c

  990 format(i5,1x,f16.8,1x,f16.8,1x,f16.8,1x,f16.8)

c..........
   99 stop
   90 print *,'MAIN:  I/O error'
      goto 99

      end

File: analyze_protein_trajectory/py_tools_monomer/average_ene_vs_residue/average_ene_vs_residue.py
import sys
import glob
import numpy as np

"""
calculate the average of ene_vs_residue

input  : path of ene_res files in thm decomposition directory
         "../../traj_A/3D-RISM/thm_decomposition/thma_data/"

output : average_ene_vs_residue.dat

"""

dat_dir = sys.argv[1]

list_files = sorted(glob.glob('{}/*/ene_res.dat'.format(dat_dir)))
ndata = len(list_files)

# read data and caluclate the statistics
data = []
ene_all = []
for k, file in enumerate(list_files):
    ene_sum = 0.

    with open(file, 'r') as f:
        lines = f.readlines()[1:]  # skip first line (num. of residues)
        for line in lines:
            line = line.replace('\n','')
            ene_res = float(line[6:23])
            ene_sum += ene_res
            data.append(line)

    ene_all.append(ene_sum)

    if k == 0:
        n_res = len(data)
        print('number of residues    =', n_res)


print('total number of data  =', ndata)
print('average_ene_all       =', np.mean(ene_all))

f_out = open('summary_average_ene_vs_residue.dat', 'w')
print('total number of data  =', ndata, file=f_out)
print('', file=f_out)
print('average_ene_all       =', np.mean(ene_all), file=f_out)
print('', file=f_out)
f_out.close()


# calculate the average of ene_vs_residue
f_out = open('average_ene_vs_residue.dat', 'w')

for i in range(1, n_res+1):
    # select data by residue number
    res_x = [x for x in data if int(x[:6]) == i]

    l_ene_res = []
    l_ene_LJ  = []
    l_ene_cou = []
    l_ene_sig = []

    for line in res_x:
        ene_res = float(line[6:23])
        ene_res_LJ = float(line[23:40])
        ene_res_cou = float(line[40:57])

        l_ene_res.append(ene_res)
        l_ene_LJ.append(ene_res_LJ)
        l_ene_cou.append(ene_res_cou)
        l_ene_sig.append(ene_res*ene_res)

    ave_ene_tot = np.mean(l_ene_res)
    ave_ene_LJ = np.mean(l_ene_LJ)
    ave_ene_cou = np.mean(l_ene_cou)

    sig_ene_tot = np.sqrt(np.mean(l_ene_sig) - ave_ene_tot**2)

    print('{:>6d} {:16.8f} {:16.8f} {:16.8f} {:16.8f}'.format(i,
                            ave_ene_tot, ave_ene_LJ, ave_ene_cou, sig_ene_tot),
                            file=f_out)

f_out.close()

# =============================================================================
